src/deploy/functions/release.ts

```diff
@@ -0,0 +1,477 @@
+        trigger: functionTrigger,
+        sourceUploadUrl: sourceUrl,
+        labels: _.assign({}, deploymentTool.labels(), functionInfo.labels),
+        availableMemoryMb: functionInfo.availableMemoryMb,
+        timeout: functionInfo.timeout,
+        runtime: runtime,
+        maxInstances: functionInfo.maxInstances,
+        vpcConnector: functionInfo.vpcConnector,
+        vpcConnectorEgressSettings: functionInfo.vpcConnectorEgressSettings,
+        serviceAccountEmail: functionInfo.serviceAccountEmail,
+        environmentVariables: _.assign(
+          {},
+          existingFunction.environmentVariables,
+          defaultEnvVariables
+        ),
+      };
+      deployments.push({
+        name: name,
+        retryFunction: async function () {
+          return cloudfunctions.updateFunction(options);
+        },
+        trigger: functionTrigger,
+      });
+    })
+    .value();
+
+  // Delete functions
+  const functionsToDelete = _.chain(existingFunctions)
+    .filter((functionInfo) => {
+      return deploymentTool.check(functionInfo.labels);
+    }) // only delete functions uploaded via firebase-tools
+    .map((fn) => {
+      return _.get(fn, "name");
+    })
+    .difference(uploadedNames)
+    .intersection(deleteReleaseNames)
+    .value();
+
+  if (functionsToDelete.length) {
+    const deleteList = _.map(functionsToDelete, (func) => {
+      return "\t" + helper.getFunctionLabel(func);
+    }).join("\n");
+
+    if (options.nonInteractive && !options.force) {
+      const deleteCommands = _.map(functionsToDelete, (func) => {
+        return (
+          "\tfirebase functions:delete " +
+          helper.getFunctionName(func) +
+          " --region " +
+          helper.getRegion(func)
+        );
+      }).join("\n");
+
+      throw new FirebaseError(
+        "The following functions are found in your project but do not exist in your local source code:\n" +
+          deleteList +
+          "\n\nAborting because deletion cannot proceed in non-interactive mode. To fix, manually delete the functions by running:\n" +
+          clc.bold(deleteCommands)
+      );
+    } else if (!options.force) {
+      logger.info(
+        "\nThe following functions are found in your project but do not exist in your local source code:\n" +
+          deleteList +
+          "\n\nIf you are renaming a function or changing its region, it is recommended that you create the new " +
+          "function first before deleting the old one to prevent event loss. For more info, visit " +
+          clc.underline("https://firebase.google.com/docs/functions/manage-functions#modify" + "\n")
+      );
+    }
+
+    let proceed = true;
+    if (!options.force) {
+      proceed = await promptOnce({
+        type: "confirm",
+        name: "confirm",
+        default: false,
+        message:
+          "Would you like to proceed with deletion? Selecting no will continue the rest of the deployments.",
+      });
+    }
+    if (proceed) {
```

Comment:
nit: If we tested for `!proceed` (which I'd probably call `skip` to avoid a double negative) we would have the smaller block sooner. Then we wouldn't forget what the `else` block nearly 100 lines lower is referring to.